distributed/worker.py

```diff
@@ -920,3 +717,666 @@ def run(worker, stream, function=None, args=(), kwargs={}):
+        if key in self.data:
+            return
+
+        self.data[key] = value
+
+        if key not in self.nbytes:
+            self.nbytes[key] = sizeof(value)
+
+        for dep in self.dependents.get(key, ()):
+            if dep in self.waiting_for_data:
+                if key in self.waiting_for_data[dep]:
+                    self.waiting_for_data[dep].remove(key)
+                if not self.waiting_for_data[dep]:
+                    self.waiting_for_data[dep]
+                    self.transition(dep, 'ready')
+
+    @gen.coroutine
+    def gather_dep(self, dep):
+        try:
+            if self.validate:
+                self.validate_state()
+
+            while True:
+                if not self.who_has.get(dep):
+                    # TODO: ask scheduler nicely for new who_has before canceling
+                    if dep not in self.dependents:
+                        return
+                    for key in list(self.dependents[dep]):
+                        if dep in self.executing:
+                            continue
+                        if dep in self.waiting_for_data.get(key, ()):
+                            self.cancel_key(key)
+                    return
+                worker = random.choice(list(self.who_has[dep]))
+                ip, port = worker.split(':')
+                try:
+                    future = connect(ip, int(port))
+                    self.connections[future] = True
+                    stream = yield gen.with_timeout(timedelta(seconds=3),
+                                                    future)
+                except AttributeError:  # set_sock_opt sometimes fails
+                    pass
+                except (gen.TimeoutError, StreamClosedError):
+                    with ignoring(KeyError):  # other coroutine may have removed
+                        for d in self.has_what.pop(worker):
+                            self.who_has[d].remove(worker)
+                else:
+                    break
+                finally:
+                    del self.connections[future]
+
+
+            if dep in self.data or dep in self.in_flight:  # someone beat us
+                stream.close() # close newly opened stream
+                return
+
+            deps = {dep}
+
+            total_bytes = self.nbytes[dep]
+            L = self.pending_data_per_worker[worker]
+
+            while L:
+                d = L.popleft()
+                if (d in self.data or
+                    d in self.in_flight or
+                    d in self.executing or
+                    d not in self.nbytes):  # no longer tracking
+                    continue
+                if total_bytes + self.nbytes[d] > self.target_message_size:
+                    break
+                deps.add(d)
+                total_bytes += self.nbytes[d]
+
+            self.connections[stream] = deps
+            for d in deps:
+                if d in self.in_flight:
+                    self.in_flight[d].add(stream)
```

Comment:
Perhaps assert it cannot happen instead? This would let such bugs appear clearly.